pkg/monitor/controller_test.go

```diff
@@ -0,0 +1,719 @@
+	cnpInformer := crdInformerFactory.Crd().V1alpha1().ClusterNetworkPolicies()
+	anpInformer := crdInformerFactory.Crd().V1alpha1().NetworkPolicies()
+	tierInformer := crdInformerFactory.Crd().V1alpha1().Tiers()
+	cgInformer := crdInformerFactory.Crd().V1alpha3().ClusterGroups()
+	grpInformer := crdInformerFactory.Crd().V1alpha3().Groups()
+	externalNodeInformer := crdInformerFactory.Crd().V1alpha1().ExternalNodes()
+
+	addressGroupStore := store.NewAddressGroupStore()
+	appliedToGroupStore := store.NewAppliedToGroupStore()
+	networkPolicyStore := store.NewNetworkPolicyStore()
+	groupStore := store.NewGroupStore()
+	groupEntityIndex := grouping.NewGroupEntityIndex()
+	labelIdentityIndex := labelidentity.NewLabelIdentityIndex()
+
+	networkPolicyController := networkpolicy.NewNetworkPolicyController(client,
+		crdClient,
+		groupEntityIndex,
+		labelIdentityIndex,
+		namespaceInformer,
+		serviceInformer,
+		networkPolicyInformer,
+		nodeInformer,
+		cnpInformer,
+		anpInformer,
+		tierInformer,
+		cgInformer,
+		grpInformer,
+		addressGroupStore,
+		appliedToGroupStore,
+		networkPolicyStore,
+		groupStore,
+		false)
+
+	controllerQuerier := querier.NewControllerQuerier(networkPolicyController, 10349)
+	externalNodeEnabled := true
+	controllerMonitor := NewControllerMonitor(crdClient, nodeInformer, externalNodeInformer, controllerQuerier, externalNodeEnabled)
+
+	return &fakeController{
+		controllerMonitor,
+		crdClient,
+		client,
+		informerFactory,
+		crdInformerFactory,
+	}
+}
+
+func initController(controller *fakeController, stopCh chan struct{}) {
+	controller.informerFactory.Start(stopCh)
+	controller.crdInformerFactory.Start(stopCh)
+	controller.informerFactory.WaitForCacheSync(stopCh)
+	controller.crdInformerFactory.WaitForCacheSync(stopCh)
+}
+
+func TestSyncExternalNode(t *testing.T) {
+	en := &v1alpha1.ExternalNode{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "TestExternalNode",
+			Namespace: "ns1",
+			Labels:    map[string]string{"en": "vm2"},
+		},
+	}
+	tc := []struct {
+		name                 string
+		key                  string
+		existingExternalNode *v1alpha1.ExternalNode
+		expectedAgentCRD     *v1beta1.AntreaAgentInfo
+		expectedError        string
+	}{
+		{
+			name:          "Invalid key format",
+			key:           "namespace/name/error",
+			expectedError: "unexpected key format: \"namespace/name/error\"",
+		},
+		{
+			name:          "Key does not exist",
+			key:           "default/vm2-e8be5",
+			expectedError: "",
+		},
+		{
+			name:                 "key exists",
+			key:                  "ns1/TestExternalNode",
+			existingExternalNode: en,
+			expectedAgentCRD: &v1beta1.AntreaAgentInfo{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:            "TestExternalNode",
+					ResourceVersion: "",
+					Generation:      0,
+				},
+				NetworkPolicyControllerInfo: v1beta1.NetworkPolicyControllerInfo{
+					NetworkPolicyNum:  0,
+					AddressGroupNum:   0,
+					AppliedToGroupNum: 0,
```

Comment:
maybe you could omit all the fields which have a default value (all the fields except for the `Name`). It will not impact the test but it will help reduce the verbosity.